treat carriage returns as white space (might solve #129)

mrbaseman opened the pull request:
carriage returns (that might have come in from copy&paste out of dos/win files) can lead to a damaged config file

DimitriPapadopoulos commented:
Ah... Good catch!

Just a short question though. The DOS/Windows new line is `\r\n` compared to `\n` on Unix. Therefore you do need to discard a trailing `\r`but why would you want to discard heading ` \r`? This should never happen in any normal situation - meaning the probability of having a heading \r` is not any larger than having other non-printable ASCII characters (NUL, EOT, BEL, VT, FF, FS, GS, ...) elsewhere in the configuration file.

mrbaseman commented:
if you cat the config file together from several pieces, e.g. a config template which happens to have dos line endings and you append some lines with `echo "something" >> config` you end up with a mixture of line endings. If you then join two lines, the `\r` may end anywhere in the file. The result probably also depends on what the editor thinks about the format of the file. If we treat the carriage returns as white space I think it makes sense to do so everywhere, just like tabs and space characters.

mrbaseman commented:
@DimitriPapadopoulos shall we check for other non-printable ASCII characters, too?

DimitriPapadopoulos commented:
I wouldn't check for other characters. Thinking about it, we've already gone beyond the call of duty: I don't think other software do that, except perhaps specialized software such as some text editors. For example _awk_ fails (very) miserably on simple text files with DOS/Windows newlines.

DimitriPapadopoulos commented:
Actually I would rather see more low-level functions remove `\r`when reading the configuration file. But that would be a much more involved change!

DimitriPapadopoulos commented:
Perhaps an _iswhitespace_-like static function or macro would help clarify the existing code.

mrbaseman commented:
you mean like [this](https://github.com/adrienverge/openfortivpn/pull/334/commits/20fc7e8587a0ac3a6e1659304826a1352220c19a)?

DimitriPapadopoulos commented:
Looks good, just remove the `#endif` though. Perhaps you should use an inline function, in modern C compilers it should be as fast a macro and actually recommended.
[6.43 An Inline Function is As Fast As a Macro](https://gcc.gnu.org/onlinedocs/gcc/Inline.html)
[PRE00-C. Prefer inline or static functions to function-like macros](https://wiki.sei.cmu.edu/confluence/pages/viewpage.action?pageId=87152416)

adrienverge commented:
It could worth adding other unicode whitespace characters as well?
http://jkorpela.fi/chars/spaces.html

DimitriPapadopoulos commented:
It's as you wish, but at I said I feel we've already gone beyond the call of duty. I don't know of any software that would go to such lengths to discard unexpected non-printable characters when reading a configuration file.

Also we should limit ourselves to ASCII. The encoding of the configuration is not defined, so it is expected to be ASCII, not UTF-8 or anything else. Do we need non-ASCII characters?

adrienverge commented:
Agreed that this is already much better as is! So it's fine to if you merge like that :+1: 

I know that strange whitespace characters can be copied-pasted in text files, sadly it happens sometimes. Should we support unicode and and detect it? Maybe not, you're right.

mrbaseman commented:
I was thinking of an array, too, or `if(character > 0) && (character <= 32))`, but I agree with you @DimitriPapadopoulos, we have already covered the most probable cases.

@adrienverge, with UTF-8 we would have to change really [a lot of things](https://www.cprogramming.com/tutorial/unicode.html) and when handling UTF-8 one should avoid implementing this on the low level, and rather use [iwwspace](http://pubs.opengroup.org/onlinepubs/9699919799/functions/iswspace.html) for instance

adrienverge commented:
OK!

DimitriPapadopoulos commented:
Not sure `const` is useful here by the way, `char` is OK (this is not a pointer). Also it just occurred to me that most _is_-like functions expect and return `int`:
https://linux.die.net/man/3/isspace

DimitriPapadopoulos commented:
Perhaps _isspace()_ is actually enough here, except it doesn't handle `\0`.

DimitriPapadopoulos commented:
Perhaps the _isspace()_ and _iscntrl()_ combination is what we need.

mrbaseman commented:
@DimitriPapadopoulos yes, you are right, the `const` is obsolete here. When calling by reference or with a pointer it would give a hint to the compiler that we may not change the value inside of the inline function. I already had merged this branch, but I can create a new PR.

Actually, `isspace()` does what we need. The additional check `character != '\0'` is obsolete anyway. If `character` is one of the space characters, it is already not `'\0'`. Note, that I have changed the code a little bit when I have introduced this function. The additional check was not there when removing the trailing white spaces, but going backwards through a null terminated string one should never hit a `'\0'` character. But also when removing the heading white spaces, as long as we find one of `' ', '\t', '\r'` it's anyway not the terminating `'\0'`. 

After this analysis, I'm going to open another pull request for switching to `isspace()`.
I'll open another pull request
